{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## MyVariant.info and MyGene.info Use Case"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The following R script demonstrates the utility of MyVariant.info and MyGene.info R clients to annotate variants and prioritize candidate genes in patients with rare Mendelian diseases. This specific study uses data obtained from the database of phenotype and genotype (dbGaP) study...FASTQ files generated by Ng et al for the Miller syndrome study. FASTQ files were processed according to the Broad Institute’s best practices. Individual samples were aligned to the hg19 reference genome using BWA-MEM 0.7.10. Variants were called using GATK 3.3-0 HaplotypeCaller and quality scores were recalibrated using GATK VariantRecalibrator."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: GenomicFeatures\n",
      "Loading required package: AnnotationDbi\n",
      "Loading required package: Biobase\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "Attaching package: ‘AnnotationDbi’\n",
      "\n",
      "The following object is masked from ‘package:dplyr’:\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "Attaching package: ‘mygene’\n",
      "\n",
      "The following object is masked from ‘package:dplyr’:\n",
      "\n",
      "    query\n",
      "\n",
      "Loading required package: DBI\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(myvariant)\n",
    "library(mygene)\n",
    "library(VariantAnnotation)\n",
    "library(GO.db)\n",
    "source(\"https://raw.githubusercontent.com/Network-of-BioThings/myvariant.info/master/docs/ipynb/mendelian.R\")\n",
    "setwd(\"~/sulab/myvariant/vcf/recal\")\n",
    "vcf.files <- paste(getwd(), list.files(getwd()), sep=\"/\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "mendelian.R defines some helper functions that are used in the analysis downstream of retrieving annotations:\n",
    "\n",
    "`replaceWith0` - replaces all NAs in a data.frame with 0.\n",
    "\n",
    "`rankByCaddScore` - for prioritizing genes by deleteriousness (scaled CADD score)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "## Annotating variants with MyVariant.info\n",
    "The following function reads in each output VCF file using the VariantAnnotation package available from Bioconductor. Install with `biocLite(\"VariantAnnotation\")`. `formatHgvs` (from the ***myvariant*** Bioconductor package) is a function that reads the genomic location and variant information from the VCF to create HGVS IDs which serve as a primary key for each variant. The function `getVariants` makes the queries to MyVariant.info to retrieve annotations."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing  /Users/Adam/sulab/myvariant/vcf/recal/subject01_recalibrate_SNP_vqsr.vcf ...\n",
      "found header lines for 3 ‘fixed’ fields: ALT, QUAL, FILTER \n",
      "found header lines for 24 ‘info’ fields: AC, AF, ..., VQSLOD, culprit \n",
      "found header lines for 5 ‘geno’ fields: GT, AD, DP, GQ, PL \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Querying chunk 1 of 37\n",
      "Querying chunk 2 of 37\n",
      "Querying chunk 3 of 37\n",
      "Querying chunk 4 of 37\n",
      "Querying chunk 5 of 37\n",
      "Querying chunk 6 of 37\n",
      "Querying chunk 7 of 37\n",
      "Querying chunk 8 of 37\n",
      "Querying chunk 9 of 37\n",
      "Querying chunk 10 of 37\n",
      "Querying chunk 11 of 37\n",
      "Querying chunk 12 of 37\n",
      "Querying chunk 13 of 37\n",
      "Querying chunk 14 of 37\n",
      "Querying chunk 15 of 37\n",
      "Querying chunk 16 of 37\n",
      "Querying chunk 17 of 37\n",
      "Querying chunk 18 of 37\n",
      "Querying chunk 19 of 37\n",
      "Querying chunk 20 of 37\n",
      "Querying chunk 21 of 37\n",
      "Querying chunk 22 of 37\n",
      "Querying chunk 23 of 37\n",
      "Querying chunk 24 of 37\n",
      "Querying chunk 25 of 37\n",
      "Querying chunk 26 of 37\n",
      "Querying chunk 27 of 37\n",
      "Querying chunk 28 of 37\n",
      "Querying chunk 29 of 37\n",
      "Querying chunk 30 of 37\n",
      "Querying chunk 31 of 37\n",
      "Querying chunk 32 of 37\n",
      "Querying chunk 33 of 37\n",
      "Querying chunk 34 of 37\n",
      "Querying chunk 35 of 37\n",
      "Querying chunk 36 of 37\n",
      "Querying chunk 37 of 37\n",
      "Concatenating data, please be patient.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing  /Users/Adam/sulab/myvariant/vcf/recal/subject02_recalibrate_SNP_vqsr.vcf ...\n",
      "found header lines for 3 ‘fixed’ fields: ALT, QUAL, FILTER \n",
      "found header lines for 24 ‘info’ fields: AC, AF, ..., VQSLOD, culprit \n",
      "found header lines for 5 ‘geno’ fields: GT, AD, DP, GQ, PL \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Querying chunk 1 of 36\n",
      "Querying chunk 2 of 36\n",
      "Querying chunk 3 of 36\n",
      "Querying chunk 4 of 36\n",
      "Querying chunk 5 of 36\n",
      "Querying chunk 6 of 36\n",
      "Querying chunk 7 of 36\n",
      "Querying chunk 8 of 36\n",
      "Querying chunk 9 of 36\n",
      "Querying chunk 10 of 36\n",
      "Querying chunk 11 of 36\n",
      "Querying chunk 12 of 36\n",
      "Querying chunk 13 of 36\n",
      "Querying chunk 14 of 36\n",
      "Querying chunk 15 of 36\n",
      "Querying chunk 16 of 36\n",
      "Querying chunk 17 of 36\n",
      "Querying chunk 18 of 36\n",
      "Querying chunk 19 of 36\n",
      "Querying chunk 20 of 36\n",
      "Querying chunk 21 of 36\n",
      "Querying chunk 22 of 36\n",
      "Querying chunk 23 of 36\n",
      "Querying chunk 24 of 36\n",
      "Querying chunk 25 of 36\n",
      "Querying chunk 26 of 36\n",
      "Querying chunk 27 of 36\n",
      "Querying chunk 28 of 36\n",
      "Querying chunk 29 of 36\n",
      "Querying chunk 30 of 36\n",
      "Querying chunk 31 of 36\n",
      "Querying chunk 32 of 36\n",
      "Querying chunk 33 of 36\n",
      "Querying chunk 34 of 36\n",
      "Querying chunk 35 of 36\n",
      "Querying chunk 36 of 36\n",
      "Concatenating data, please be patient.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing  /Users/Adam/sulab/myvariant/vcf/recal/subject03_recalibrate_SNP_vqsr.vcf ...\n",
      "found header lines for 3 ‘fixed’ fields: ALT, QUAL, FILTER \n",
      "found header lines for 24 ‘info’ fields: AC, AF, ..., VQSLOD, culprit \n",
      "found header lines for 5 ‘geno’ fields: GT, AD, DP, GQ, PL \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Querying chunk 1 of 29\n",
      "Querying chunk 2 of 29\n",
      "Querying chunk 3 of 29\n",
      "Querying chunk 4 of 29\n",
      "Querying chunk 5 of 29\n",
      "Querying chunk 6 of 29\n",
      "Querying chunk 7 of 29\n",
      "Querying chunk 8 of 29\n",
      "Querying chunk 9 of 29\n",
      "Querying chunk 10 of 29\n",
      "Querying chunk 11 of 29\n",
      "Querying chunk 12 of 29\n",
      "Querying chunk 13 of 29\n",
      "Querying chunk 14 of 29\n",
      "Querying chunk 15 of 29\n",
      "Querying chunk 16 of 29\n",
      "Querying chunk 17 of 29\n",
      "Querying chunk 18 of 29\n",
      "Querying chunk 19 of 29\n",
      "Querying chunk 20 of 29\n",
      "Querying chunk 21 of 29\n",
      "Querying chunk 22 of 29\n",
      "Querying chunk 23 of 29\n",
      "Querying chunk 24 of 29\n",
      "Querying chunk 25 of 29\n",
      "Querying chunk 26 of 29\n",
      "Querying chunk 27 of 29\n",
      "Querying chunk 28 of 29\n",
      "Querying chunk 29 of 29\n",
      "Concatenating data, please be patient.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing  /Users/Adam/sulab/myvariant/vcf/recal/subject04_recalibrate_SNP_vqsr.vcf ...\n",
      "found header lines for 3 ‘fixed’ fields: ALT, QUAL, FILTER \n",
      "found header lines for 24 ‘info’ fields: AC, AF, ..., VQSLOD, culprit \n",
      "found header lines for 5 ‘geno’ fields: GT, AD, DP, GQ, PL \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Querying chunk 1 of 35\n",
      "Querying chunk 2 of 35\n",
      "Querying chunk 3 of 35\n",
      "Querying chunk 4 of 35\n",
      "Querying chunk 5 of 35\n",
      "Querying chunk 6 of 35\n",
      "Querying chunk 7 of 35\n",
      "Querying chunk 8 of 35\n",
      "Querying chunk 9 of 35\n",
      "Querying chunk 10 of 35\n",
      "Querying chunk 11 of 35\n",
      "Querying chunk 12 of 35\n",
      "Querying chunk 13 of 35\n",
      "Querying chunk 14 of 35\n",
      "Querying chunk 15 of 35\n",
      "Querying chunk 16 of 35\n",
      "Querying chunk 17 of 35\n",
      "Querying chunk 18 of 35\n",
      "Querying chunk 19 of 35\n",
      "Querying chunk 20 of 35\n",
      "Querying chunk 21 of 35\n",
      "Querying chunk 22 of 35\n",
      "Querying chunk 23 of 35\n",
      "Querying chunk 24 of 35\n",
      "Querying chunk 25 of 35\n",
      "Querying chunk 26 of 35\n",
      "Querying chunk 27 of 35\n",
      "Querying chunk 28 of 35\n",
      "Querying chunk 29 of 35\n",
      "Querying chunk 30 of 35\n",
      "Querying chunk 31 of 35\n",
      "Querying chunk 32 of 35\n",
      "Querying chunk 33 of 35\n",
      "Querying chunk 34 of 35\n",
      "Querying chunk 35 of 35\n",
      "Concatenating data, please be patient.\n"
     ]
    }
   ],
   "source": [
    "getVars <- function(vcf.file){\n",
    "  cat(paste(\"Processing \", vcf.file, \"...\\n\", sep=\" \"))\n",
    "  vcf <- readVcf(vcf.file, genome=\"hg19\")\n",
    "  vcf <- vcf[isSNV(vcf)]\n",
    "  vars <- rowRanges(vcf)\n",
    "  vars <- as(vars, \"DataFrame\")\n",
    "  vars$query <- formatHgvs(vcf, \"snp\")\n",
    "  annotations <- getVariants(vars$query, fields=c(\"dbnsfp.genename\", \"dbnsfp.1000gp1.af\", \n",
    "                                                  \"exac.af\", \"cadd.consequence\", \"cadd.phred\"))\n",
    "  annotations[c('DP', 'FS', 'QD')] <- info(vcf)[c('DP', 'FS', 'QD')]\n",
    "  annotations <- replaceWith0(annotations)\n",
    "  annotations\n",
    "}\n",
    "\n",
    "vars <- lapply(vcf.files, getVars)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We then filter each of the VCF files based on annotations."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2442 common genes are mutated amongst patients"
     ]
    }
   ],
   "source": [
    "cat(paste(nrow(subset(data.frame(table(unlist(lapply(vars, function(i) unique(i$dbnsfp.genename))))), \n",
    "    Freq == 4)), \"common genes are mutated amongst patients\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "## Filtering Steps\n",
    "The next steps apply annotation based filters by subsetting each data.frame by those variants that meet each criteria.\n",
    "\n",
    "1) The first filter utilizes annotations output by HaplotypeCaller from GATK. Here we keep variants with at least 8 reads, Fisher Strand bias less than 30, and quality over depth greater than 2. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2309 genes remain after filtering for coverage and strand bias"
     ]
    }
   ],
   "source": [
    "filter1 <- lapply(vars, function(i) subset(i, DP > 8 & FS < 30 & QD > 2))\n",
    "    \n",
    "cat(paste(nrow(subset(data.frame(table(unlist(lapply(filter1, function(i) unique(i$dbnsfp.genename))))), \n",
    "    Freq == 4)), \"genes remain after filtering for coverage and strand bias\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "2) Mendelian diseases are most likely to be caused by nonsynonymous mutations. We can take advantage of the CADD database that denotes the mutation type in the field \"cadd.consequence\"."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1918 genes remain after filtering for nonsynonymous and splice site variants"
     ]
    }
   ],
   "source": [
    "filter2 <- lapply(filter1, function(i) subset(i, cadd.consequence %in% c(\"NON_SYNONYMOUS\", \"STOP_GAINED\", \"STOP_LOST\", \n",
    "                                           \"CANONICAL_SPLICE\", \"SPLICE_SITE\")))\n",
    "    \n",
    "cat(paste(nrow(subset(data.frame(table(unlist(lapply(filter2, function(i) unique(i$dbnsfp.genename))))), \n",
    "    Freq == 4)), \"genes remain after filtering for nonsynonymous and splice site variants\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "3) The third filter keeps rare variants according to the ExAC data set with allele frequency < 0.01. Rare diseases are likely caused by mutations that have not been documented yet."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19 genes remain after filtering for allele frequency < 0.01 in the ExAC dataset"
     ]
    }
   ],
   "source": [
    "filter3 <- lapply(filter2, function(i) subset(i, exac.af < 0.01))\n",
    "    \n",
    "cat(paste(nrow(subset(data.frame(table(unlist(lapply(filter3, function(i) unique(i$dbnsfp.genename))))), \n",
    "    Freq == 4)), \"genes remain after filtering for allele frequency < 0.01 in the ExAC dataset\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "4) The last filter keeps rare variants according to the 1000 Genomes Project with allele frequency < 0.01."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9 genes remain after filtering for allele frequency < 0.01 in 1000 Genomes Project"
     ]
    }
   ],
   "source": [
    "filter4 <- lapply(filter3, function(i) subset(i, sapply(dbnsfp.1000gp1.af, function(j) j < 0.01 )))\n",
    "    \n",
    "top.genes <- subset(data.frame(table(unlist(lapply(filter4, function(i) unique(i$dbnsfp.genename))))), Freq == 4)\n",
    "top.genes <- subset(top.genes, !(Var1 %in% c(\"NULL\", 0)))\n",
    "cat(paste(nrow(top.genes), \"genes remain after filtering for allele frequency < 0.01 in 1000 Genomes Project\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<ol class=list-inline>\n",
       "\t<li>CDC27</li>\n",
       "\t<li>CDON</li>\n",
       "\t<li>CTBP2</li>\n",
       "\t<li>DHODH</li>\n",
       "\t<li>FAM104B</li>\n",
       "\t<li>HYDIN</li>\n",
       "\t<li>KIR3DL3</li>\n",
       "\t<li>PIK3R3</li>\n",
       "\t<li>XKR3</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item CDC27\n",
       "\\item CDON\n",
       "\\item CTBP2\n",
       "\\item DHODH\n",
       "\\item FAM104B\n",
       "\\item HYDIN\n",
       "\\item KIR3DL3\n",
       "\\item PIK3R3\n",
       "\\item XKR3\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. CDC27\n",
       "2. CDON\n",
       "3. CTBP2\n",
       "4. DHODH\n",
       "5. FAM104B\n",
       "6. HYDIN\n",
       "7. KIR3DL3\n",
       "8. PIK3R3\n",
       "9. XKR3\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] CDC27   CDON    CTBP2   DHODH   FAM104B HYDIN   KIR3DL3 PIK3R3  XKR3   \n",
       "968 Levels: 0 ABCA2 ABCA6 ABCB6 ABCC10 ABCC11 ABCC5 ABHD5 ABLIM3 ABTB1 ... ZWINT"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "top.genes$Var1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "## Prioritizing genes\n",
    "We can prioritize the genes that contain variants in each of the patients according to CADD (deleteriousness) score. `rankByCaddScore` extracts the average CADD scores of the variants in each gene per patient and ranks in descending order."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "In is.na(e1) | is.na(e2): longer object length is not a multiple of shorter object lengthWarning message:\n",
      "In `!=.default`(gene, c(\"NULL\", 0)): longer object length is not a multiple of shorter object length"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<thead><tr><th></th><th scope=col>gene</th><th scope=col>cadd.phred</th></tr></thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>DHODH</td><td>26.81</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>CTBP2</td><td>21.385</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>PIK3R3</td><td>20.7</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>HYDIN</td><td>18.91</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>CDC27</td><td>18.545</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>XKR3</td><td>15.72</td></tr>\n",
       "\t<tr><th scope=row>7</th><td>CDON</td><td>10.02</td></tr>\n",
       "\t<tr><th scope=row>8</th><td>FAM104B</td><td>7.2055</td></tr>\n",
       "\t<tr><th scope=row>9</th><td>KIR3DL3</td><td>1.012</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "\\begin{tabular}{r|ll}\n",
       "  & gene & cadd.phred\\\\\n",
       "\\hline\n",
       "\t1 & DHODH & 26.81\\\\\n",
       "\t2 & CTBP2 & 21.385\\\\\n",
       "\t3 & PIK3R3 & 20.7\\\\\n",
       "\t4 & HYDIN & 18.91\\\\\n",
       "\t5 & CDC27 & 18.545\\\\\n",
       "\t6 & XKR3 & 15.72\\\\\n",
       "\t7 & CDON & 10.02\\\\\n",
       "\t8 & FAM104B & 7.2055\\\\\n",
       "\t9 & KIR3DL3 & 1.012\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/plain": [
       "     gene cadd.phred\n",
       "1   DHODH    26.8100\n",
       "2   CTBP2    21.3850\n",
       "3  PIK3R3    20.7000\n",
       "4   HYDIN    18.9100\n",
       "5   CDC27    18.5450\n",
       "6    XKR3    15.7200\n",
       "7    CDON    10.0200\n",
       "8 FAM104B     7.2055\n",
       "9 KIR3DL3     1.0120"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ranked <- rankByCaddScore(top.genes$Var1, filter4)\n",
    "ranked"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "##Annotating candidate genes with MyGene.info\n",
    "Now we can annotate the genes using the `queryMany` function from ***mygene*** which can be install from Bioconductor (`biocLite(\"mygene\")`). We specify `scopes=symbol` and request annotations by the `field` parameter. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Finished\n"
     ]
    }
   ],
   "source": [
    "res <- data.frame(queryMany(ranked$gene, scopes=\"symbol\", species=\"human\", fields=c(\"go.BP\", \"name\", \"MIM\", \"uniprot\")))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[[1]]\n",
       "                                                     term         id evidence\n",
       "1                 pyrimidine nucleobase metabolic process GO:0006206      TAS\n",
       "2    'de novo' pyrimidine nucleobase biosynthetic process GO:0006207      IBA\n",
       "3                                        female pregnancy GO:0007565      IEA\n",
       "4                                               lactation GO:0007595      IEA\n",
       "5                                    response to caffeine GO:0031000      IEA\n",
       "6                                        response to drug GO:0042493      IEA\n",
       "7                                  response to starvation GO:0042594      IEA\n",
       "8                positive regulation of apoptotic process GO:0043065      IEA\n",
       "9                      'de novo' UMP biosynthetic process GO:0044205      IEA\n",
       "10                       small molecule metabolic process GO:0044281      TAS\n",
       "11             pyrimidine nucleoside biosynthetic process GO:0046134      TAS\n",
       "12 nucleobase-containing small molecule metabolic process GO:0055086      TAS\n",
       "13                            oxidation-reduction process GO:0055114      IEA\n",
       "14                    regulation of mitochondrial fission GO:0090140      IEA\n",
       "\n",
       "[[2]]\n",
       "                                                                   term\n",
       "1  negative regulation of transcription from RNA polymerase II promoter\n",
       "2                                          transcription, DNA-templated\n",
       "3                             negative regulation of cell proliferation\n",
       "4                                              viral genome replication\n",
       "5                              positive regulation of chromatin binding\n",
       "6                   negative regulation of transcription, DNA-templated\n",
       "7  positive regulation of transcription from RNA polymerase II promoter\n",
       "8       positive regulation of retinoic acid receptor signaling pathway\n",
       "9                                        white fat cell differentiation\n",
       "10                                          oxidation-reduction process\n",
       "           id evidence   pubmed\n",
       "1  GO:0000122      IBA       NA\n",
       "2  GO:0006351      IEA       NA\n",
       "3  GO:0008285      TAS  9479502\n",
       "4  GO:0019079      TAS  9479502\n",
       "5  GO:0035563      IEA       NA\n",
       "6  GO:0045892      ISS       NA\n",
       "7  GO:0045944      IMP 23775127\n",
       "8  GO:0048386      IMP 23775127\n",
       "9  GO:0050872      ISS       NA\n",
       "10 GO:0055114      IEA       NA\n",
       "\n",
       "[[3]]\n",
       "                                                                   term\n",
       "1                                        phospholipid metabolic process\n",
       "2                             phosphatidylinositol biosynthetic process\n",
       "3                                    insulin receptor signaling pathway\n",
       "4                 phosphatidylinositol-3-phosphate biosynthetic process\n",
       "5                  regulation of phosphatidylinositol 3-kinase activity\n",
       "6                                      small molecule metabolic process\n",
       "7                                  phosphatidylinositol phosphorylation\n",
       "8 positive regulation of glucose import in response to insulin stimulus\n",
       "          id evidence  pubmed\n",
       "1 GO:0006644      TAS      NA\n",
       "2 GO:0006661      TAS      NA\n",
       "3 GO:0008286      TAS 9524259\n",
       "4 GO:0036092      TAS 9524259\n",
       "5 GO:0043551      IBA      NA\n",
       "6 GO:0044281      TAS      NA\n",
       "7 GO:0046854      IBA      NA\n",
       "8 GO:2001275      ISS      NA\n",
       "\n",
       "[[4]]\n",
       "                            term         id evidence\n",
       "1    epithelial cell development GO:0002064      IEA\n",
       "2                cilium movement GO:0003341      IEA\n",
       "3 ventricular system development GO:0021591      IEA\n",
       "4                cilium assembly GO:0042384      IEA\n",
       "5            trachea development GO:0060438      IEA\n",
       "\n",
       "[[5]]\n",
       "                                                                                                              term\n",
       "1                                                                                               mitotic cell cycle\n",
       "2                                                              metaphase/anaphase transition of mitotic cell cycle\n",
       "3                                                                              mitotic spindle assembly checkpoint\n",
       "4                                                                                               cell proliferation\n",
       "5                   anaphase-promoting complex-dependent proteasomal ubiquitin-dependent protein catabolic process\n",
       "6                          negative regulation of ubiquitin-protein ligase activity involved in mitotic cell cycle\n",
       "7 positive regulation of ubiquitin-protein ligase activity involved in regulation of mitotic cell cycle transition\n",
       "8                                   regulation of ubiquitin-protein ligase activity involved in mitotic cell cycle\n",
       "9                                                                                protein K11-linked ubiquitination\n",
       "          id evidence   pubmed\n",
       "1 GO:0000278      TAS       NA\n",
       "2 GO:0007091      IMP  7736578\n",
       "3 GO:0007094      TAS       NA\n",
       "4 GO:0008283      TAS  7736578\n",
       "5 GO:0031145      TAS       NA\n",
       "6 GO:0051436      TAS       NA\n",
       "7 GO:0051437      TAS       NA\n",
       "8 GO:0051439      TAS       NA\n",
       "9 GO:0070979      IDA 18485873\n",
       "\n",
       "[[6]]\n",
       "NULL\n",
       "\n",
       "[[7]]\n",
       "                                                                   term\n",
       "1                                               cell fate specification\n",
       "2                                   lens development in camera-type eye\n",
       "3                                                         cell adhesion\n",
       "4                                                       myoblast fusion\n",
       "5                              anterior/posterior pattern specification\n",
       "6                                          embryonic body morphogenesis\n",
       "7                        skeletal muscle satellite cell differentiation\n",
       "8                                           cerebral cortex development\n",
       "9                                           muscle cell differentiation\n",
       "10                                  positive regulation of MAPK cascade\n",
       "11                    regulation of protein heterodimerization activity\n",
       "12                        positive regulation of neuron differentiation\n",
       "13 positive regulation of transcription from RNA polymerase II promoter\n",
       "14            positive regulation of skeletal muscle tissue development\n",
       "15     positive regulation of small GTPase mediated signal transduction\n",
       "16                   positive regulation of muscle cell differentiation\n",
       "17                    embryonic retina morphogenesis in camera-type eye\n",
       "18           positive regulation of neural precursor cell proliferation\n",
       "           id evidence  pubmed\n",
       "1  GO:0001708      IEA      NA\n",
       "2  GO:0002088      IEA      NA\n",
       "3  GO:0007155      TAS 9214393\n",
       "4  GO:0007520      IEA      NA\n",
       "5  GO:0009952      IEA      NA\n",
       "6  GO:0010172      IEA      NA\n",
       "7  GO:0014816      IEA      NA\n",
       "8  GO:0021987      IEA      NA\n",
       "9  GO:0042692      TAS      NA\n",
       "10 GO:0043410      IEA      NA\n",
       "11 GO:0043497      IEA      NA\n",
       "12 GO:0045666      IEA      NA\n",
       "13 GO:0045944      IEA      NA\n",
       "14 GO:0048643      IEA      NA\n",
       "15 GO:0051057      IEA      NA\n",
       "16 GO:0051149      TAS      NA\n",
       "17 GO:0060059      IEA      NA\n",
       "18 GO:2000179      IEA      NA\n",
       "\n",
       "[[8]]\n",
       "NULL\n",
       "\n",
       "[[9]]\n",
       "NULL\n"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res$go.BP"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "# GO annotation\n",
    "We can then use the Bioconductor package ***GO.db*** to map GO IDs to GO:0008152, the biological process id for metabolic process."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<ol class=list-inline>\n",
       "\t<li>DHODH</li>\n",
       "\t<li>CTBP2</li>\n",
       "\t<li>PIK3R3</li>\n",
       "\t<li>CDC27</li>\n",
       "\t<li>CDON</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item DHODH\n",
       "\\item CTBP2\n",
       "\\item PIK3R3\n",
       "\\item CDC27\n",
       "\\item CDON\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. DHODH\n",
       "2. CTBP2\n",
       "3. PIK3R3\n",
       "4. CDC27\n",
       "5. CDON\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] DHODH  CTBP2  PIK3R3 CDC27  CDON  \n",
       "Levels: CDC27 CDON CTBP2 DHODH FAM104B HYDIN KIR3DL3 PIK3R3 XKR3"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "miller.bp <- lapply(res$go.BP, function(i) unlist(i$id))\n",
    "bp.ancestor <- lapply(miller.bp, function(i) sapply(i, function(j) \"GO:0008152\" %in% unlist(GOBPANCESTOR[[j]])))\n",
    "candidate.genes <- ranked$gene[sapply(bp.ancestor, function(i) TRUE %in% i)]\n",
    "candidate.genes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "The final candidate gene list contains only genes involved in metabolic processes where perterbation is more likely to result in mendelian disease. Through the use of MyVariant.info and MyGene.info annotation services, we have narrowed the candidate gene list from 2589 genes to seven. While we cannot fully remove the process of biological interpretation from an analysis, we have greatly reduced the burden."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.2.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
